pkg/winperfcounters/scraper.go

```diff
@@ -0,0 +1,157 @@
+
+import (
+	"fmt"
+
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters/internal/pdh"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/winperfcounters/internal/third_party/telegraf/win_perf_counters"
+)
+
+var _ PerfCounterWatcher = (*Watcher)(nil)
+
+// PerfCounterWatcher represents how to scrape data
+type PerfCounterWatcher interface {
+	// Path returns the counter path
+	Path() string
+	// ScrapeData collects a measurement and returns the value(s).
+	ScrapeData() ([]win_perf_counters.CounterValue, error)
+	// Close all counters/handles related to the query and free all associated memory.
+	Close() error
+
+	GetMetricRep() MetricRep
+}
+
+const instanceLabelName = "instance"
+
+type WatcherCfg struct {
+	ObjectCfg ObjectConfig
+}
+
+type MetricRep struct {
+	Name       string
+	Attributes map[string]string
+}
+
+type Watcher struct {
+	Counter *pdh.PerfCounter
+	MetricRep
+}
+
+func (w Watcher) Path() string {
+	return w.Counter.Path()
+}
+
+func (w Watcher) ScrapeData() ([]win_perf_counters.CounterValue, error) {
+	return w.Counter.ScrapeData()
+}
+
+func (w Watcher) Close() error {
+	return w.Counter.Close()
+}
+
+func (w Watcher) GetMetricRep() MetricRep {
+	return w.MetricRep
+}
+
+// BuildPaths creates watchers and their paths from configs.
+func BuildPaths(scraperCfgs []WatcherCfg) ([]PerfCounterWatcher, error) {
+	var errs error
+	var watchers []PerfCounterWatcher
+
+	for _, scraperCfg := range scraperCfgs {
+		for _, instance := range scraperCfg.ObjectCfg.instances() {
+			for _, counterCfg := range scraperCfg.ObjectCfg.Counters {
+				counterPath := counterPath(scraperCfg.ObjectCfg.Object, instance, counterCfg.Name)
+
+				c, err := pdh.NewPerfCounter(counterPath, true)
+				if err != nil {
+					errs = multierr.Append(errs, fmt.Errorf("counter %v: %w", counterPath, err))
+				} else {
+					newWatcher := Watcher{Counter: c}
+
+					if counterCfg.Metric != "" {
+						metricCfg := MetricRep{Name: counterCfg.Metric}
+						if counterCfg.Attributes != nil {
+							metricCfg.Attributes = counterCfg.Attributes
+						}
+						newWatcher.MetricRep = metricCfg
+					} else {
+						newWatcher.MetricRep.Name = c.Path()
+					}
+
+					watchers = append(watchers, newWatcher)
+				}
+			}
+		}
+	}
+
+	return watchers, errs
+}
+
+func counterPath(object, instance, counterName string) string {
+	if instance != "" {
+		instance = fmt.Sprintf("(%s)", instance)
+	}
+
+	return fmt.Sprintf("\\%s%s\\%s", object, instance, counterName)
+}
+
+type CounterValue struct {
+	MetricRep
+	Value int64
+}
+
+// WatchCounters pulls values given the passed in watchers
+func WatchCounters(watchers []PerfCounterWatcher) (metrics []CounterValue, errs error) {
+	for _, scraper := range watchers {
+		counterValues, err := scraper.ScrapeData()
+		if err != nil {
+			errs = multierr.Append(errs, err)
+			continue
+		}
+		metric := scraper.GetMetricRep()
```

Comment:
no because it should be a single value pulled from the countervalues based on how we build the paths. This does bring up the point be can probably get away with just `counterValues[0]`.

Reply:
Updated